pkg/controller/rbac/provider/binding/reconciler.go

```diff
@@ -0,0 +1,218 @@
+type ReconcilerOption func(*Reconciler)
+
+// WithLogger specifies how the Reconciler should log messages.
+func WithLogger(log logging.Logger) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.log = log
+	}
+}
+
+// WithRecorder specifies how the Reconciler should record Kubernetes events.
+func WithRecorder(er event.Recorder) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.record = er
+	}
+}
+
+// WithClientApplicator specifies how the Reconciler should interact with the
+// Kubernetes API.
+func WithClientApplicator(ca resource.ClientApplicator) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.client = ca
+	}
+}
+
+// NewReconciler returns a Reconciler of ProviderRevisions.
+func NewReconciler(mgr manager.Manager, opts ...ReconcilerOption) *Reconciler {
+	kube := unstructured.NewClient(mgr.GetClient())
+
+	r := &Reconciler{
+		mgr: mgr,
+
+		// TODO(negz): Is Updating appropriate here? Probably.
+		client: resource.ClientApplicator{
+			Client:     kube,
+			Applicator: resource.NewAPIUpdatingApplicator(kube),
+		},
+
+		log:    logging.NewNopLogger(),
+		record: event.NewNopRecorder(),
+	}
+
+	for _, f := range opts {
+		f(r)
+	}
+	return r
+}
+
+// A Reconciler reconciles ProviderRevisions.
+type Reconciler struct {
+	client resource.ClientApplicator
+	mgr    manager.Manager
+
+	log    logging.Logger
+	record event.Recorder
+}
+
+// Reconcile a ProviderRevision by creating a ClusterRoleBinding that binds a
+// provider's service account to its system ClusterRole.
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	pr := &v1alpha1.ProviderRevision{}
+	if err := r.client.Get(ctx, req.NamespacedName, pr); err != nil {
+		// In case object is not found, most likely the object was deleted and
+		// then disappeared while the event was in the processing queue. We
+		// don't need to take any action in that case.
+		log.Debug(errGetPR, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetPR)
+	}
+
+	log = log.WithValues(
+		"uid", pr.GetUID(),
+		"version", pr.GetResourceVersion(),
+		"name", pr.GetName(),
+	)
+
+	if meta.WasDeleted(pr) {
+		// There's nothing to do if our PR is being deleted. Any ClusterRoles
+		// we created will be garbage collected by Kubernetes.
+		return reconcile.Result{Requeue: false}, nil
+	}
+
+	l := &corev1.ServiceAccountList{}
+	if err := r.client.List(ctx, l); err != nil {
+		log.Debug(errListSAs, "error", err)
+		r.record.Event(pr, event.Warning(reasonBind, errors.Wrap(err, errListSAs)))
+		return reconcile.Result{RequeueAfter: shortWait}, nil
+	}
+
+	// Filter down to the ServiceAccounts that are owned by this
+	// ProviderRevision. Each revision should control at most one, but it's easy
+	// and relatively harmless for us to handle there being many.
```

Comment:
Currently we are naming the SA the same name as the revision and they all exist in a single namespace for a given Crossplane deployment. Do you anticipate needing this more robust strategy for a future scenario?

Reply:
Not really - it just seemed more future proof to discover the SA rather than rely on assumptions about how other controllers behave. If listing and filtering all SAs proves to be an issue I'd be fine with changing the behaviour.